src/StreamJsonRpc/SystemTextJsonFormatter.cs

```diff
@@ -0,0 +1,1414 @@
+            }
+
+            JsonElement? valueElement = null;
+            switch (this.JsonArguments?.ValueKind)
+            {
+                case JsonValueKind.Object when name is not null:
+                    if (this.JsonArguments.Value.TryGetProperty(name, out JsonElement propertyValue))
+                    {
+                        valueElement = propertyValue;
+                    }
+
+                    break;
+                case JsonValueKind.Array:
+                    int elementIndex = 0;
+                    foreach (JsonElement arrayElement in this.JsonArguments.Value.EnumerateArray())
+                    {
+                        if (elementIndex++ == position)
+                        {
+                            valueElement = arrayElement;
+                            break;
+                        }
+                    }
+
+                    break;
+                default:
+                    throw new JsonException("Unexpected value kind for arguments: " + (this.JsonArguments?.ValueKind.ToString() ?? "null"));
+            }
+
+            try
+            {
+                using (this.formatter.TrackDeserialization(this))
+                {
+                    try
+                    {
+                        value = valueElement?.Deserialize(typeHint ?? typeof(object), this.formatter.massagedUserDataSerializerOptions);
+                    }
+                    catch (Exception ex)
+                    {
+                        if (this.formatter.JsonRpc?.TraceSource.Switch.ShouldTrace(TraceEventType.Warning) ?? false)
+                        {
+                            this.formatter.JsonRpc.TraceSource.TraceEvent(TraceEventType.Warning, (int)JsonRpc.TraceEvents.MethodArgumentDeserializationFailure, Resources.FailureDeserializingRpcArgument, name, position, typeHint, ex);
+                        }
+
+                        throw new RpcArgumentDeserializationException(name, position, typeHint, ex);
+                    }
+                }
+            }
+            catch (JsonException ex)
+            {
+                throw new RpcArgumentDeserializationException(name, position, typeHint, ex);
+            }
+
+            return valueElement.HasValue;
+        }
+
+        protected override TopLevelPropertyBagBase? CreateTopLevelPropertyBag() => new TopLevelPropertyBag(this.formatter.massagedUserDataSerializerOptions);
+
+        protected override void ReleaseBuffers()
+        {
+            base.ReleaseBuffers();
+            this.jsonArguments = null;
+            this.formatter.deserializingDocument?.Dispose();
+            this.formatter.deserializingDocument = null;
+        }
+
+        private static int CountArguments(JsonElement arguments)
+        {
+            int count = 0;
+            switch (arguments.ValueKind)
+            {
+                case JsonValueKind.Array:
+                    foreach (JsonElement element in arguments.EnumerateArray())
```

Comment:
Could this be used here: https://learn.microsoft.com/en-us/dotnet/api/system.text.json.jsonelement.getarraylength?view=net-8.0 to avoid enumerating?

Reply:
or JsonElement.Item.Length?

Reply:
Yes. Excellent.